observe/datadog.go

```diff
@@ -0,0 +1,100 @@
+package observe
+
+import (
+	datadog "github.com/DataDog/opencensus-go-exporter-datadog"
+	"github.com/kelseyhightower/envconfig"
+	"github.com/pkg/errors"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/trace"
+)
+
+// DatadogExporterConfig provides configuration for the Datadog exporter.
+// Config can be initiated with envconfig from environment
+type DatadogExporterConfig struct {
+	DatadogExporterEnabled        bool   `default:"false" split_words:"true"`
+	DatadogExporterMetricsAddress string `split_words:"true"`
+	DatadogExporterTracesAddress  string `split_words:"true"`
+	DatadogExporterNamespace      string `default:"opencensus" split_words:"true"`
+}
+
+// RegisterAndObserveDatadog will initiate and register Datadog metrics
+// and traces exporter when enabled
+func RegisterAndObserveDatadog(onError func(error)) error {
+	if SkipObserve() {
+		return nil
+	}
+
+	var config DatadogExporterConfig
+	envconfig.Process("", &config)
+
+	exp, err := NewDatadogExporter(config, onError)
+	if err != nil {
+		return errors.Wrap(err, "unable to initiate Datadog's opencensus exporter")
+	}
+
+	if exp != nil {
+		trace.RegisterExporter(exp)
+		view.RegisterExporter(exp)
+	}
+
+	return nil
+}
+
+// NewDatadogExporter will return Datadog's opencensus exporter if it's enabled (through
+// DATADOG_EXPORTER_ENABLED env variable). When the exporter is disabled, it will
+// return nil. Exporter will send metrics and traces to Datadog's agent using
+// addresses specified through DatadogExporterConfig
+func NewDatadogExporter(config DatadogExporterConfig, onErr func(error)) (*datadog.Exporter, error) {
+
+	if !config.DatadogExporterEnabled {
```

Comment:
I think this should be done in line 29, not here. IMO if you call `NewDatadogExporter()` you should already have amde decision if you want DD enabled or not. For the caller, if `err` is nil then exp should be usable.

Reply:
At least at the beginning, I'd like to give people a way to control if the DD exporter is enabled through that flag, even if they have it implemented (e.g. enable/disable in different environments).

I'm OK with moving this check to line 29, but we would need to make this constructor private, so that we have a single entry that evaluated the flag. What do you think about it?

Reply:
Why would this New function need to be not exported?

Reply:
If we leave the New function public and move the `enable` check outside of it, users will have to implement exactly the thing that the `RegisterAndObserveDatadog` is already doing. And I really think we should allow users to control if the individual exporters are enabled through flags, rather than implemented or not.

Having both, `RegisterAndObserveDatadog` and the `New...`, public, would force the users to make a choice which one to use. It is not necessary and we can save folks from spending time on decisions they don't have to make.

If you think it would be better (maybe for testing?) to keep the `New...` one, I would remove the `RegisterAndObserveDatadog` and keep the enable flag check in the constructor.

Also, if we don't want to return `nil` if the exporter is disabled, we should go back to returning error from the function.

@oliver-nyt  What do you think?

Reply:
I disagree. `NewDatadogExporter` should give you a new datadog exporter (as long as you provide a valid config). If you call this is up to you. If the caller uses `RegisterAndObserveDatadog` then they get the beenfot of the checks and config options you mentioned. If the caller uses  `NewDatadogExporter` then they won't - which seems fine to me.

As for ` we should go back to returning error from the function.` - fine with me.

Reply:
@oliver-nyt Thanks for the feedback. Let's drop the `RegisterAndObserveDatadog`, so that we can avoid what I'm concerned about and make sure the package behaves the way you described (which I agree with).

I updated the package and the implementation in kitserver.

Reply:
👍